### ISSUE TITLE
cursor moves to the beginning of the line while typing quickly if the CodeMirror is controlled and the render of the component containing CodeMirror takes too long

### ISSUE THREAD

**Issue by andreimatei** [uiwjs/react-codemirror#694]
Hello! I believe I'm running into a bug. My symptoms are similar to the old https://github.com/uiwjs/react-codemirror/issues/199, but I'm using the latest version (`4.23.6`) and also the workaround from https://github.com/uiwjs/react-codemirror/issues/199#issuecomment-941943581 does not work for me.

So, I'm typing into my controlled `<CodeMirror>` and, every few keystrokes, the cursor jumps to the beginning of the text box. I believe triggering the issue is dependent on `<CodeMirror>` being used in a "heavy" component: if the parent of `<CodeMirror>` (which supplies the `value`) is simple, then I don't get the cursor jump. If the parent component has a bunch of other sub-components (and so the render takes longer), then I do get the bug -- so I suspect the issue is some sort of race condition whereby an update of `value` arrives "too late", after more key presses.

My `CodeMirror` invocation looks like this:
```
    const [value, setValue] = useState("");
    <CodeMirror
      value={value}
      onChange={(value) => setValue(value)}
    />
    
    <MyOtherHeavyComponent/>
```

If I remove `<MyOtherHeavyComponent/>`, then it all works fine. Also, if I type slowly, it again works fine.

**Comment by azizghuloum** [comment]
Yeah.  This is pretty bad.  I now have to type code character by character otherwise the cursor jumps to the beginning and messes everything up.

**Comment by adamnieto** [comment]
This helped me but not sure if it will help you. So I was looking at [this issue](https://github.com/uiwjs/react-codemirror/issues/700) and I modified the code here to also call the user provided `onChange` and it stopped moving the cursor to the beginning.

**Comment by xiao252** [comment]
> This helped me but not sure if it will help you. So I was looking at [this issue](https://github.com/uiwjs/react-codemirror/issues/700) and I modified the code here to also call the user provided `onChange` and it stopped moving the cursor to the beginning.这对我有帮助，但不确定是否对你有帮助。所以我正在研究[这个问题](https://github.com/uiwjs/react-codemirror/issues/700)，我修改了这里的代码以也调用用户提供的`onChange`并且它停止将光标移动到开头。

"Use setValue(val) only when you want React to re-render the component". This still has problems for me, the cursor runs to the starting position

**Comment by seansps** [comment]
This is still an issue in `4.23.12` -- has anyone identified a workaround? I've tried what was suggested in #700 with no luck.

**Comment by jaywcjlove** [comment]
@nsrCodes You're very welcome to submit a PR for optimization!

This `useEffect` is mainly used to replace the editor content with the new `value` when the external `value` changes.
The line `annotations: [External.of(true)]` is typically used to mark the update as externally triggered, so that plugins or extensions can recognize and handle it accordingly.

Looking forward to your improvements! 🙌

https://github.com/uiwjs/react-codemirror/blob/154727c817478f85d267005639ab390dbf9bde6f/core/src/useCodeMirror.ts#L162-L173

**Comment by jaywcjlove** [comment]
@nsrCodes I tested your PR and reviewed the code carefully; I didn’t find any obvious issues. I’ll keep an eye on the potential performance impact of `setInterval(..., 1)`.
I plan to merge the PR and release a new version.

Thank you for your contribution!
